core/src/sumeragi/fault.rs

```diff
@@ -310,991 +403,465 @@ pub fn run_sumeragi_main_loop<F>(
+                    _ => {
+                        trace!("Leader not handling message, {:?}", msg);
+                    }
                 }
+            } else {
+                should_sleep = true;
+            }
 
-                if voting_block_option.is_none() {
-                    let transactions = sumeragi.queue.get_transactions_for_block(&wsv_guard);
-                    if transactions.len() == 0 {
-                        continue;
-                    }
-                    println!(
-                        "I, {}, have transactions to make a block with.",
-                        sumeragi.peer_id.public_key
-                    );
+            if voting_block_option.is_none() {
+                if state_machine_guard.transaction_cache.len() == 0 {
+                    instant_when_we_should_create_a_block = Instant::now() + sumeragi.block_time;
+                    continue;
+                }
+                if Instant::now() > instant_when_we_should_create_a_block
+                    || state_machine_guard.transaction_cache.len() >= sumeragi.queue.txs_in_block
+                {
+                    let transactions: Vec<VersionedAcceptedTransaction> = state_machine_guard
+                        .transaction_cache
+                        .iter()
+                        .map(|tx| tx.clone().unwrap())
+                        .collect();
+
+                    info!("sumeragi Doing block with {} txs.", transactions.len());
                     // TODO: This should properly process triggers
                     let event_recommendations = Vec::new();
 
                     let block = PendingBlock::new(transactions, event_recommendations).chain(
                         state_machine_guard.latest_block_height,
                         state_machine_guard.latest_block_hash,
-                        ViewChangeProofs::empty(), // self.view_change_proofs().clone(),
-                        Vec::new(),                //self.invalidated_blocks_hashes.clone(),
+                        view_change_proof_chain.clone(),
                     );
                     {
-                        let block = block.validate(&sumeragi.transaction_validator, &wsv_guard);
+                        let block = {
+                            let span_for_sumeragi_leader_block_validate =
+                                span!(Level::TRACE, "Sumeragi Leader Create block, validation.");
+                            let _enter_for_sumeragi_leader_block_validate =
+                                span_for_sumeragi_leader_block_validate.enter();
+
+                            let block = block.validate(
+                                &sumeragi.transaction_validator,
+                                &state_machine_guard.wsv,
+                            );
+                            block
+                        };
 
                         for event in Vec::<Event>::from(&block) {
                             trace!(?event);
```

Comment:
We had discussion about avoiding calling `drop`, i think maybe this one pop up after rebase.

Reply:
yes